File: JAVA/TraceGPS/src/classes/PasserelleXML.java
// Projet TraceGPS - API Java
// Fichier : Passerelle.java
// Cette classe abstraite fournit les outils permettant d'obtenir un document XML à partir d'un fichier ou d'un service web
// Dernière mise à jour : 26/3/2018 par Jim

package classes;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;

import java.net.HttpURLConnection;
import java.net.URL;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;

public abstract class PasserelleXML {

    // méthode protégée statique pour obtenir un flux en lecture (java.io.InputStream)
    // à partir de l'adresse d'un fichier ou de l'URL d'un service web
    protected static InputStream getFluxEnLecture(String adrFichierOuServiceWeb)
    {
		InputStream unFluxEnLecture;
		try
		{
			if (adrFichierOuServiceWeb.startsWith("http"))
			{	// connexion HTTP au service web
				URL url = new URL(adrFichierOuServiceWeb);
				HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();

				// récupération de la réponse dans un flux en lecture (InputStream)
				unFluxEnLecture = new BufferedInputStream(urlConnection.getInputStream());

			}
			else
			{	// création d'un flux en lecture (InputStream) depuis le fichier
				unFluxEnLecture = new FileInputStream(new File(adrFichierOuServiceWeb));
			}
			return unFluxEnLecture;
		}
		catch (Exception ex)
		{	return null;
		}	
    }

    // méthode protégée statique pour obtenir document XML (org.w3c.dom.Document)
    // à partir d'un flux de données en lecture (java.io.InputStream)
	protected static Document getDocumentXML(InputStream unFluxEnLecture)
	{
		try
		{
			// création d'une instance de DocumentBuilderFactory et DocumentBuilder
			DocumentBuilderFactory leDBF = DocumentBuilderFactory.newInstance();
			DocumentBuilder leDB = leDBF.newDocumentBuilder();
	
			// on crée un nouveau document XML avec en argument le flux XML
			Document leDocument = leDB.parse(unFluxEnLecture);
			return leDocument;
		}
		catch (Exception ex)
		{	return null;
		}	
	}
}

File: C#/TraceGPS_C#_fourni/TraceGPS/modele/PasserelleXML.cs
// Projet TraceGPS
// fichier : modele/PasserelleXML.cs
// Rôle : Cette classe abstraite fixe les outils pour "parser" un fichier GPS et mettre à jour un objet Trace
// Dernière mise à jour : 1/11/2021 par dp

using System;
using System.Net;
using System.IO;
using System.Xml;				// permet d'utiliser les classes XML

namespace TraceGPS
{
    public abstract class PasserelleXML
    {
        // méthode protégée statique pour obtenir un flux en lecture à partir de l'adresse d'un fichier ou de l'URL d'un service web
        // paramètre adrFichierOuServiceWeb : le nom du fichier contenant la trace (String)
        // retourne : un flux de données en lecture (System.IO.StreamReader)
        protected static StreamReader getFluxEnLecture(String adrFichierOuServiceWeb)
        {
            StreamReader unFluxEnLecture;
            if (adrFichierOuServiceWeb.StartsWith("http"))
            {   // l'adresse fournie est l'URL d'un service web car elle commence par "http"
                // création d'une requête http
                HttpWebRequest uneRequeteHttp = (HttpWebRequest)WebRequest.Create(adrFichierOuServiceWeb);
                uneRequeteHttp.Method = WebRequestMethods.Http.Get;
                // récupération de la réponse
                WebResponse uneReponseHttp = uneRequeteHttp.GetResponse();
                // création d'un flux en lecture (SteamReader) à partir de la réponse web
                unFluxEnLecture = new StreamReader(uneReponseHttp.GetResponseStream());
            }
            else
            {   // l'adresse fournie est celle d'un fichier
                // création d'un flux en lecture (StreamReader) depuis le fichier
                unFluxEnLecture = File.OpenText(adrFichierOuServiceWeb);
            }
            return unFluxEnLecture;
        }

        // méthode protégée statique pour obtenir un document XML à partir d'un flux de données en lecture
        // paramètre unFluxEnLecture : un flux de données en lecture (System.IO.StreamReader)
        // retourne : un document XML (System.Xml.Document)
        protected static XmlReader getDocumentXML(StreamReader unFluxEnLecture)
        {
            try
            {
                XmlReader leDocument = XmlReader.Create(unFluxEnLecture);
                return leDocument;
            }
            catch (Exception ex)
            {
                return null;
            }
        }
    }
}
